src/app/client-view/training-view/training-day/exercise/exercise.component.html

```diff
@@ -61,11 +61,11 @@ <h5 class="card-title">Completed Work</h5>
           </div>
         </li>
       </ul>
-      <div
+      <!-- <div
         class="d-flex justify-content-end"
         *ngIf="this.trainingForm.get('weight').valid && this.trainingForm.get('RPE').valid">E1RM for {{ exercise.exerciseName }}:
```

Comment:
In the HTML you don't need `this`, it's assumed by Angular
```suggestion
        *ngIf="trainingForm.get('weight').valid && trainingForm.get('RPE').valid">E1RM for {{ exercise.exerciseName }}:
```

I think this needs more complex logic associated with it, for example another variable in the component TS file that tracks whether the form has been submitted or whether a pair of inputs are valid, because currently there could be many `weight`s and many `RPE`s and it doesn't know which to pick. Then you need so set the max variable at some time when it would be valid, probably when the form has been saved again? Then we can uncomment this section.

Reply:
If I tag this logic in the `onSave` method in the exercise component is it not meant to work?
`  onSave(): void {
    // console.log(this.trainingForm.get('setForms').value);
    this.exercise.completedTraining = this.trainingForm.get('setForms').value;
    this.trainingService.updateServerBlock(this.weekNo, this.dayNo, this.exercise);
    if (this.trainingForm.get('weight').valid && this.trainingForm.get('RPE').valid) {
      this.showRM = true;
    }
    else {this.showRM = false;}
  }`
I assumed that the `this.trainingForm.get('weight')` and `this.trainingForm.get('RPE')` correspond to whichever the field of inputs the user presses the input from...?

Reply:
The problem is that `trainingForm` doesn't directly have `weight` or `RPE` parameters. The hierarchy looks like this:

1.  `trainingForm`

2. `setForms`

3.  a number of elements that all have: 

4. `weight`, `RPE`, `notes`

So to access the `weight`, `RPE`, `notes` we need to go via the `setForms` array (which can be done with `get()`) and then we can pick out the `value` and that will contain the data we need.

Reply:
Otherwise the logic is fine, although `showRM` should be set to false initially when it is declared, too.

Reply:
I tried to unbpack the `setForms` array using this:
`for (const [key, value] of Object.entries(this.trainingForm.get('setForms').value)) {
      console.log(`${key}: ${value}`);
    }` in the `onSave` method but I'm returned an `[object Object]` in the console... How can I pick out specific entries in the setForms array?

Reply:
Use 
`for (const form of this.trainingForm.get('setForms').value) {
...
}`

`setForms.value` is an array, so it doesn't have key-value pairs. It contains a number of objects which do have key value pairs - you can just access them with the `.` operator.
If you're implementing this in the `onSave()` method, `this.trainingForm.get('setForms').value` is already stored as `this.exercise.completedTraining` so it would be more efficient to use that variable rather than calling the `get` method. Using `get` would be a more general way of implementing it.

Reply:
`console.log(this.exercise.completedTraining.weight);` doesn't work?

Btw I've turned the RPE input into a dropdown with options from @ 5 to @ 10, and added a css class for that invalid case.